### ISSUE TITLE
Multiplexing doesn't increase throughput on the same socket

### ISSUE THREAD

**Issue by buybackoff** [aspnet/WebSockets#122]
I am trying to get the max throughput and cannot get any expected speedup from multiplexing on a server. On a client, multiplexing over the same `ManagedWebSocket` increased throughput by almost 3x (from 12 Kops to 32 Kops), but multiplexing on the server has no effect, even somewhat negative.

The server code is below. The client is doing the reverse, with the same `Task.Run` for parallel send/receive.

```
private async Task EchoSimple(WebSocket webSocket) {
    byte[] buffer = new byte[64];
    var result = await webSocket.ReceiveAsync(new ArraySegment<byte>(buffer), CancellationToken.None);
    while (!result.CloseStatus.HasValue) {
        await webSocket.SendAsync(new ArraySegment<byte>(buffer, 0, result.Count), result.MessageType, result.EndOfMessage, CancellationToken.None);
        result = await webSocket.ReceiveAsync(new ArraySegment<byte>(buffer), CancellationToken.None);
    }
    await webSocket.CloseAsync(result.CloseStatus.Value, result.CloseStatusDescription, CancellationToken.None);
}

private async Task EchoMultiplex(WebSocket webSocket) {
    // from https://github.com/alexandrnikitin/MPMCQueue.NET
    // SCC.ConcurrentQueue killed performance by 5x, PerfView showed 30% work only for it
    var queue = new MPMCQueue(128);
    var cts = new CancellationTokenSource();
    var t = Task.Run(async () => {
        while (!cts.IsCancellationRequested) {
            object buffer;
            if (queue.TryDequeue(out buffer)) {
                await webSocket.SendAsync(new ArraySegment<byte>((byte[])buffer), WebSocketMessageType.Binary, true, CancellationToken.None);
            }
        }
    }, cts.Token);
    byte[] buffer2 = new byte[64];
    var result = await webSocket.ReceiveAsync(new ArraySegment<byte>(buffer2), CancellationToken.None);
    while (!result.CloseStatus.HasValue) {
        while (!queue.TryEnqueue(buffer2)) { }
        buffer2 = new byte[64];
        result = await webSocket.ReceiveAsync(new ArraySegment<byte>(buffer2), CancellationToken.None);
    }
    await webSocket.CloseAsync(result.CloseStatus.Value, result.CloseStatusDescription, CancellationToken.None);
    cts.Cancel();
}
```

With the `EchoSimple` and multiplexed client the throughput is c.32Kops, with `EchoMultiplex` it is c.30Kops. With `SCC.ConcurrentQueue` performance drops to c.6Kops and a profiler shows that the queue does a lot of work.

Given the very close results of the two versions I am suspicious that there is some implicit lock that prevents simultaneous read and write on the server? Could you please advice if that is the case and if there is a way to get the expected throughput speedup?

I'm using Kestrel 1.0.1 and Microsoft.AspNetCore.WebSockets.Server 0.1.0.

**Update:**

Maybe my understanding was incorrect - if a client sends and receives in parallel, by the time the server queues a message it is simultaneously receiving another message on some IO thread. So when I call ReceiveAsync, the next message is already in some buffer and is returned immediately. If that is the case, then c.30Kops in the upper throughput limit on my machine and cannot be increased by manual multiplexing, because the server is already doing IO in parallel. But it looks quite low on localhost.

**Comment by muratg** [comment]
We are closing this issue because no further action is planned for this issue. If you still have any issues or questions, please log a new issue with any additional details that you have.
